File: union-find/union-find.cpp
#include <algorithm>

// union-find.cpp
// author: Seong Yong-ju <sei40kr@gmail.com>

// Union-Find 木の実装
// cf https://www.slideshare.net/chokudai/union-find-49066733
const int MAX_N = 2014;

int par[MAX_N];

void init(int n) {
  std::fill(par, par + n, -1);
}

int root(int x) {
  return par[x] == x ? x : par[x] = root(par[x]);
}

bool same(int x, int y) {
  return root(x) == root(y);
}

void unite(int x, int y) {
  x = root(x);
  y = root(y);
  if (x != y) {
    par[x] = y;
  }
}

File: union-find/union-find.rs
// union-find.rs
// author: Seong Yong-ju <sei40kr@gmail.com>

#[allow(dead_code)]
struct UnionFind {
    par: Vec<usize>,
    rank: Vec<usize>,
}

#[allow(dead_code)]
impl UnionFind {
    fn new(n: usize) -> UnionFind {
        UnionFind {
            par: (0..n).collect(),
            rank: vec![0; n],
        }
    }

    fn root(&mut self, x: usize) -> usize {
        if x == self.par[x] {
            x
        } else {
            let par = self.par[x];
            let new_par = self.root(par);
            self.par[x] = new_par;
            new_par
        }
    }

    fn same(&mut self, x: usize, y: usize) -> bool {
        self.root(x) == self.root(y)
    }

    fn unite(&mut self, mut x: usize, mut y: usize) {
        x = self.root(x);
        y = self.root(y);
        if x != y {
            self.par[x] = y;
        }
    }
}
